clang-tools-extra/test/clang-tidy/checkers/bugprone/capturing-this-by-field.cpp

```diff
@@ -0,0 +1,144 @@
+// RUN: %check_clang_tidy -std=c++11-or-later %s bugprone-capturing-this-by-field %t -- -config="{CheckOptions: {bugprone-capturing-this-by-field.FunctionWrapperTypes: '::std::function;::Fn'}}" --
```

Comment:
Could you please provide all the same tests, but for `std::bind` instead of lambda?

Reply:
something like this:
```
struct Basic {
  Basic() : Captured(std::bind(&Basic::perform, this)) {}
  // CHECK-MESSAGES: :[[@LINE-1]]:22: warning: using bind expressions to capture 'this' and storing it in class member
  void perform() { static_cast<void>(this); }
  std::function<void()> Captured;
  // CHECK-MESSAGES: :[[@LINE-1]]:25: note: 'std::function<void (void)>' that stores captured 'this'
};
```
Bear in mind that there is a lot variants for default bind calls - `std::bind`, `boost::bind`, `std::bind_front`, `std::bind_back`, ...

Reply:
Besides C++11 lambda and `std::bind`, there also well-known library [Boost Lambda 2](https://www.boost.org/doc/libs/1_84_0/libs/lambda2/doc/html/lambda2.html). It's also able to capture `this` and then store it inside `std::function`:
```
struct test
{
    int x = 0;
    std::function<bool(int)> fun = x==_1;
};
```
See the full snippet here: https://godbolt.org/z/8W4rrqh1s

Besides Boost Lambda 2 there also [Boost Lambda](https://www.boost.org/doc/libs/1_84_0/doc/html/lambda.html) and [Boost Phoenix](https://www.boost.org/doc/libs/1_84_0/libs/phoenix/doc/html/index.html) and ??? but I never used them and can't say anythind about them.

Reply:
@PiotrZSL @5chmidti @EugeneZelenko @vbvictor what do you think about lambdas from Boost library? Does it worth to handle in this check, or that lambdas deserve a separate check(maybe even in "boost" module)?

Reply:
Unfortunately, I am not familiar with boost library and `std::bind`. Maybe I can place FIXME and create new issue to track it.

Reply:
Oke, don't worry about bind and Boost's lambdas. I will create issues about all of them later.

Reply:
Created issues about bind and Boost's lambdas:
https://github.com/llvm/llvm-project/issues/131220
https://github.com/llvm/llvm-project/issues/131231